### ISSUE TITLE
is exists way to flush all dirty data to hdd,keep cache read performance?

### ISSUE THREAD

**Issue by perrynzhou** [Open-CAS/open-cas-linux#394]
@mmichal10  @imjfckm  we found used  opencas as glusterfs brick,we store billions of small files, that case read performance too slow.we don't know why?

```
[root@szdpl1543 ~]# casadm -L
type    id   disk       status    write policy   device
cache   1    /dev/sda   Running   wt             -
+core   1    /dev/sdd   Active    -              /dev/cas1-1
cache   2    /dev/sdb   Running   wt             -
+core   1    /dev/sde   Active    -              /dev/cas2-1

[root@szdpl1543 ~]#  casadm -P -i 1
Cache Id                  1
Cache Size                926351414 [4KiB Blocks] / 3533.75 [GiB]
Cache Device              /dev/sda
Core Devices              1
Inactive Core Devices     0
Write Policy              wt
Eviction Policy           lru
Cleaning Policy           alru
Promotion Policy          always
Cache line size           4 [KiB]
Metadata Memory Footprint 46.7 [GiB]
Dirty for                 0 [s] / Cache clean
Metadata Mode             normal
Status                    Running

+==================+===========+======+=============+
| Usage statistics |   Count   |  %   |   Units     |
+==================+===========+======+=============+
| Occupancy        | 255341184 | 27.6 | 4KiB blocks |
| Free             | 671010230 | 72.4 | 4KiB blocks |
| Clean            | 255341184 | 27.6 | 4KiB blocks |
| Dirty            |         0 |  0.0 | 4KiB blocks |
+==================+===========+======+=============+

+======================+============+=======+==========+
| Request statistics   |   Count    |   %   | Units    |
+======================+============+=======+==========+
| Read hits            | 2019477483 |  85.1 | Requests |
| Read partial misses  |          0 |   0.0 | Requests |
| Read full misses     |         64 |   0.0 | Requests |
| Read total           | 2019477547 |  85.1 | Requests |
+----------------------+------------+-------+----------+
| Write hits           |  315700875 |  13.3 | Requests |
| Write partial misses |      67257 |   0.0 | Requests |
| Write full misses    |   38355043 |   1.6 | Requests |
| Write total          |  354123175 |  14.9 | Requests |
+----------------------+------------+-------+----------+
| Pass-Through reads   |          0 |   0.0 | Requests |
| Pass-Through writes  |          0 |   0.0 | Requests |
| Serviced requests    | 2373600722 | 100.0 | Requests |
+----------------------+------------+-------+----------+
| Total requests       | 2373600722 | 100.0 | Requests |
+======================+============+=======+==========+

+==================================+=============+=======+=============+
| Block statistics                 |    Count    |   %   |   Units     |
+==================================+=============+=======+=============+
| Reads from core(s)               |         356 |   0.0 | 4KiB blocks |
| Writes to core(s)                |  1270413020 | 100.0 | 4KiB blocks |
| Total to/from core(s)            |  1270413376 | 100.0 | 4KiB blocks |
+----------------------------------+-------------+-------+-------------+
| Reads from cache                 |           0 |   0.0 | 4KiB blocks |
| Writes to cache                  |           0 |   0.0 | 4KiB blocks |
| Total to/from cache              |           0 |   0.0 | 4KiB blocks |
+----------------------------------+-------------+-------+-------------+
| Reads from exported object(s)    | 10420265964 |  89.1 | 4KiB blocks |
| Writes to exported object(s)     |  1270413020 |  10.9 | 4KiB blocks |
| Total to/from exported object(s) | 11690678984 | 100.0 | 4KiB blocks |
+==================================+=============+=======+=============+

+====================+=======+=====+==========+
| Error statistics   | Count |  %  | Units    |
+====================+=======+=====+==========+
| Cache read errors  |     0 | 0.0 | Requests |
| Cache write errors |     0 | 0.0 | Requests |
| Cache total errors |     0 | 0.0 | Requests |
+--------------------+-------+-----+----------+
| Core read errors   |     0 | 0.0 | Requests |
| Core write errors  |     0 | 0.0 | Requests |
| Core total errors  |     0 | 0.0 | Requests |
+--------------------+-------+-----+----------+
| Total errors       |     0 | 0.0 | Requests |
+====================+=======+=====+==========+
[root@szdpl1543 ~]#  casadm -P -i 2
Cache Id                  2
Cache Size                926351414 [4KiB Blocks] / 3533.75 [GiB]
Cache Device              /dev/sdb
Core Devices              1
Inactive Core Devices     0
Write Policy              wt
Eviction Policy           lru
Cleaning Policy           alru
Promotion Policy          always
Cache line size           4 [KiB]
Metadata Memory Footprint 46.7 [GiB]
Dirty for                 0 [s] / Cache clean
Metadata Mode             normal
Status                    Running

+==================+===========+=======+=============+
| Usage statistics |   Count   |   %   |   Units     |
+==================+===========+=======+=============+
| Occupancy        | 926351396 | 100.0 | 4KiB blocks |
| Free             |        18 |   0.0 | 4KiB blocks |
| Clean            | 926351396 | 100.0 | 4KiB blocks |
| Dirty            |         0 |   0.0 | 4KiB blocks |
+==================+===========+=======+=============+

+======================+============+=======+==========+
| Request statistics   |   Count    |   %   | Units    |
+======================+============+=======+==========+
| Read hits            | 3492138477 |  88.4 | Requests |
| Read partial misses  |     119602 |   0.0 | Requests |
| Read full misses     |  149293501 |   3.8 | Requests |
| Read total           | 3641551580 |  92.2 | Requests |
+----------------------+------------+-------+----------+
| Write hits           |  234420300 |   5.9 | Requests |
| Write partial misses |     155636 |   0.0 | Requests |
| Write full misses    |   32119147 |   0.8 | Requests |
| Write total          |  266695083 |   6.8 | Requests |
+----------------------+------------+-------+----------+
| Pass-Through reads   |   15353364 |   0.4 | Requests |
| Pass-Through writes  |   25311673 |   0.6 | Requests |
| Serviced requests    | 3908246663 |  99.0 | Requests |
+----------------------+------------+-------+----------+
| Total requests       | 3948911700 | 100.0 | Requests |
+======================+============+=======+==========+

+==================================+=============+=======+=============+
| Block statistics                 |    Count    |   %   |   Units     |
+==================================+=============+=======+=============+
| Reads from core(s)               |  5278071521 |  58.1 | 4KiB blocks |
| Writes to core(s)                |  3802289226 |  41.9 | 4KiB blocks |
| Total to/from core(s)            |  9080360747 | 100.0 | 4KiB blocks |
+----------------------------------+-------------+-------+-------------+
| Reads from cache                 |           0 |   0.0 | 4KiB blocks |
| Writes to cache                  |           0 |   0.0 | 4KiB blocks |
| Total to/from cache              |           0 |   0.0 | 4KiB blocks |
+----------------------------------+-------------+-------+-------------+
| Reads from exported object(s)    | 69921272402 |  94.8 | 4KiB blocks |
| Writes to exported object(s)     |  3802289226 |   5.2 | 4KiB blocks |
| Total to/from exported object(s) | 73723561628 | 100.0 | 4KiB blocks |
+==================================+=============+=======+=============+

+====================+=======+=====+==========+
| Error statistics   | Count |  %  | Units    |
+====================+=======+=====+==========+
| Cache read errors  |     0 | 0.0 | Requests |
| Cache write errors |     0 | 0.0 | Requests |
| Cache total errors |     0 | 0.0 | Requests |
+--------------------+-------+-----+----------+
| Core read errors   |     0 | 0.0 | Requests |
| Core write errors  |     0 | 0.0 | Requests |
| Core total errors  |     0 | 0.0 | Requests |
+--------------------+-------+-----+----------+
| Total errors       |     0 | 0.0 | Requests |
+====================+=======+=====+==========+
[root@szdpl1543 ~]# casadm --version
+=========================+===================+
| Name                    |      Version      |
+=========================+===================+
| CAS Cache Kernel Module | 19.09.00.00000116 |
| CAS Disk Kernel Module  | 19.09.00.00000116 |
| CAS CLI Utility         | 19.09.00.00000117 |
+=========================+===================+
[root@szdpl1543 ~]# uname -a
Linux szdpl1543 3.10.0-957.el7.x86_64 #1 SMP Thu Nov 8 23:39:32 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by mmichal10** [comment]
Hi @perrynzhou, 

In issue titile you mentioned dirtry data flush however both CAS instances you showed in logs are running in write-through mode.  With your current configuration all data is written simultaneously to cache and to backend storage and there is no dirty data to flush. 

Does the read performance degradation occurs only when you write to cas device? Are you able to determine how big is degradation? Are reads as slow as they would be issued straight to backend storage device skipping cache?

Another idea we came up with is that you are maybe using Open CAS version with performance bug. In previous issue (#342) you reported @imjfckm mentioned  d822a1d commit, which intorduced serious performance bug. Bug was fixed few days later and if there is any possibility that you are using broken version, our recommendation is to start using latest release version v20.3 (fb78466).
